closed/adds/jdk/src/share/native/jdk/crypto/jniprovider/NativeCrypto.c

```diff
@@ -318,16 +354,86 @@ JNIEXPORT jint JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
         (NULL == OSSL_RSA_private_encrypt) ||
         (NULL == OSSL_BN_bin2bn) ||
         (NULL == OSSL_BN_set_negative) ||
-        (NULL == OSSL_BN_free)) {
-        //fprintf(stderr, "One or more of the required symbols are missing in the crypto library\n");
-        //fflush(stderr);
+        (NULL == OSSL_BN_free) ||
+        ((NULL == OSSL_CRYPTO_num_locks) && (0 == ossl_ver)) ||
+        ((NULL == OSSL_CRYPTO_THREADID_set_numeric) && (0 == ossl_ver)) ||
+        ((NULL == OSSL_OPENSSL_malloc) && (0 == ossl_ver)) ||
+        ((NULL == OSSL_CRYPTO_THREADID_set_callback) && (0 == ossl_ver)) ||
+        ((NULL == OSSL_CRYPTO_set_locking_callback) && (0 == ossl_ver))) {
+        // fprintf(stderr, "One or more of the required symbols are missing in the crypto library\n");
+        // fflush(stderr);
         unload_crypto_library(handle);
         return -1;
     } else {
+        if (0 == ossl_ver) {
+            if (0 != thread_setup()) {
+                unload_crypto_library(handle);
+                return -1;
+            }
+        }
         return 0;
     }
  }
 
+#ifdef WINDOWS
+static HANDLE *lock_cs;
+
+int thread_setup()
+{
+    int i;
+    lock_cs = (*OSSL_OPENSSL_malloc)((*OSSL_CRYPTO_num_locks)() * sizeof(HANDLE));
+    if (NULL == lock_cs) {
+        return -1;
+    }
+    for (i = 0; i < (*OSSL_CRYPTO_num_locks)(); i++) {
+        lock_cs[i] = CreateMutex(NULL, FALSE, NULL);
+    }
+    (*OSSL_CRYPTO_set_locking_callback)((void (*)(int, int, char *, int))
```

Comment:
Why is there no call to `(*OSSL_CRYPTO_THREADID_set_callback)()` here like there is for non-Windows platforms?

Reply:
because for windows platform, OpenSSL already has an implementation to get thread id. 
in file `crypto/cryptlib.c` 

the code:
```
void CRYPTO_set_id_callback(unsigned long (*func) (void))
{
    id_callback = func;
}

unsigned long CRYPTO_thread_id(void)
{
    unsigned long ret = 0;

    if (id_callback == NULL) {
# ifdef OPENSSL_SYS_WIN16
        ret = (unsigned long)GetCurrentTask();
# elif defined(OPENSSL_SYS_WIN32)
        ret = (unsigned long)GetCurrentThreadId();
# elif defined(GETPID_IS_MEANINGLESS)
        ret = 1L;
# elif defined(OPENSSL_SYS_BEOS)
        ret = (unsigned long)find_thread(NULL);
# else
        ret = (unsigned long)getpid();
# endif
    } else
        ret = id_callback();
    return (ret);
}
```

Reply:
A comment about this would be welcome.